Add the list of contributors agreeing to LGPL licensing

marcusmueller opened the pull request:
* List of contributors, with short explanation
* added LGPLv3 license text as COPYING-LGPL
* moved GPLv3 license text to COPYING-GPL
* COPYING still contains the GPLv3, but with an explanation that new
  code is LGPL
* updated the contribution guide
* Removed the GPL license header from Readme, added explanation.

marcusmueller committed:
Add the list of contributors agreeing to LGPL licensing

* List of contributors, with short explanation
* added LGPLv3 license text as COPYING-LGPL
* moved GPLv3 license text to COPYING-GPL
* COPYING still contains the GPLv3, but with an explanation that new
  code is LGPL
* updated the contribution guide
* Removed the GPL license header from Readme, added explanation.

Signed-off-by: Marcus Müller <marcus@hostalia.de>

marcusmueller committed:
Agreeing to LGPL

Signed-off-by: Marcus Müller <marcus@hostalia.de>

jdemel commented on AUTHORS_GRANTING_LGPL_LICENSE.txt:
```diff
@@ -0,0 +1,16 @@
+VOLK is going to migrating from GPLv3 (GNU General Public license version 3.0)
+to LGPLv3 (GNU Lesser General Public License version 3.0).
```
It should be "or later" in both cases. For GPL that should be the case anyways and for LGPL it should be too IMHO.

michaelld replied:
actually in this case I think Marcus is correct: just GPLv3 and LGPLv3 ... but I'm open to arguments as to why one way or the other

jdemel replied:
This is the license statement in most files.

https://github.com/gnuradio/volk/blob/237a6fc9242ea8c48d2bbd417a6ea14feaf7314a/kernels/volk/volk_32fc_s32fc_multiply_32fc.h#L7-L10

I suggest to use SPDX names.
[GNU General Public License v3.0 or later (GPL-3.0-or-later)](https://spdx.org/licenses/GPL-3.0-or-later.html)
[GNU Lesser General Public License v3.0 or later (LGPL-3.0-or-later)](https://spdx.org/licenses/LGPL-3.0-or-later.html)

jdemel commented on AUTHORS_GRANTING_LGPL_LICENSE.txt:
```diff
@@ -0,0 +1,16 @@
+licensed differently (GPLv3), this gives the right to use the current
+contributions under both LGPLv3 and that other license. Future contributions by
+these authors are, however, licensed under LGPLv3, unless explicitly stated
+otherwise.
```
It should probably say "recontribute" or similar. According to the discussions we had in the corresponding GREP. @bhilburn  What's your insight on this?

michaelld replied:
I'll admit I haven't seen this part of the GREP. @jdemel can you do a quick summary of where "recontribute" comes in there? It's really not obvious to me ... what's written looks fine (with minor corrections)!

jdemel replied:
[Line 64: "All contributors need to re-submit their code under the new license"](https://github.com/gnuradio/greps/pull/33/files)

I assume the correct version would state: "contributors need to re-submit their code". And by adding their name to the list, they do so.

jdemel commented on AUTHORS_GRANTING_LGPL_LICENSE.txt:
```diff
@@ -0,0 +1,16 @@
+
+|       Date | Author (as used in commits) |
+|------------+-----------------------------|
+| 2021-07-01 | Marcus Müller               |
```
I suggest to add email and github handle. These are the infos other projects gathered. Also, this would fix a few corner cases. I have a root <root@nXXXX...> in mind.

jdemel reviewed (commented):
It looks good so far. Just had a few comments.

michaelld commented on CONTRIBUTING.md:
```diff
@@ -21,10 +21,10 @@ code.
-open source repository. It is therefore imperative that code submissions belong
-to the authors, and that submitters have the authority to merge that code into
-the public VOLK codebase.
+Any code contributions going into VOLK will become part of an LPGPL-licensed
```
Typo: "LPGPL" -> "LGPL"

michaelld commented on README.md:
```diff
@@ -98,23 +98,9 @@ We want to make sure VOLK compiles on a wide variety of compilers. Thus, we targ
+GNU Lesser General Public License version 3.0 (LGPLv3). At this point in time,
+much of the code in the repository is still GPL-licensed, but new contributors
+are asked to use the LGPLv3 for their code contributions. Existing contributors
+are very kindly requested to also allow LPGL-licensing by adding their name to
```
"their name" -> "their name, github handle, and email address used for VOLK commits -- possibly more than 1 entry if they submitted code from different handles and/or emails --" ... or something like this

jdemel replied:
I would suggest to update the license statement to LGPLv3+ and add your further explanations below. We can remove the explanations once the process is complete.

michaelld commented on AUTHORS_GRANTING_LGPL_LICENSE.txt:
```diff
@@ -0,0 +1,16 @@
+
+This file is a list of the authors who agreed to grant an LGPL license to the
+code they contributed to this repository. In case the affected code is currently
+licensed differently (GPLv3), this gives the right to use the current
```
"licensed differently (GPLv3)" -> "licensed differently than VOLK's current license (GPLv3)"

jdemel replied:
I assume the interesting part in our license is 
https://github.com/gnuradio/volk/blob/237a6fc9242ea8c48d2bbd417a6ea14feaf7314a/COPYING#L563

Followed by our license statement:
https://github.com/gnuradio/volk/blob/main/README.md#license
> VOLK is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3, or (at your option) any later version.

With these things in mind, I thought VOLK is currently licensed under GPLv3+ (or later) and I'd assume we want to re-license to LGPLv3+.

Finally, the [GREP](https://github.com/gnuradio/greps/pull/33/files) currently states:

> The proposed new license is [LGPLv3+](https://www.gnu.org/licenses/lgpl-3.0.html).

We should be consistent. I'd argue that's the most important part.

michaelld reviewed (changes requested):
good start!

jdemel commented:
@michaelld from https://github.com/gnuradio/greps/pull/33/files around line 64

> ### All contributors need to re-submit their code under the new license
>
> Whether one submitted code under a CLA or with a DCO, every contributor must re-submit their code under the new license.
> A contributor may submit their code to any project, including VOLK under a new license, even if that code was submitted under a CLA.

I assume the correct phrasing would be "I re-submit my code ..."

jdemel commented on COPYING:
```diff
@@ -1,3 +1,13 @@
+Files in this code repository are to be licensed under the LGPLv3, which you'll
```
LGPLv3+

jdemel commented on COPYING:
```diff
@@ -1,3 +1,13 @@
+find in the file COPYING-LGPL
+
+However, you'll find some files that carry a license header that assigns them as
+GPLv3, which was the default license for VOLK up to version 3. You can find the
```
They should all state GPLv3+. I hope.

marcusmueller replied:
Well, that should be easy to check?

```
cd volk/kernels/volk
ag -L "GNU General" $(ag -l copyright)
```

yields a list of only two files that have a copyright statement, but aren't GPL, and these two happen to be already LGPL.

There's a bunch of kernel, orc and asm code that doesn't carry a proper copyright note (see attached 
[unlicensed.txt](https://github.com/gnuradio/volk/files/6759814/unlicensed.txt)), but this is the exhaustive list of their committers, found by `( for f in $(ag -L license) ; do; git log  --format="* %aN <%aE>" $f ; done ) |sort -u `: 

* Albin Stigo <albin.stigo@gmail.com>
* Doug Geiger <doug.geiger@bioradiation.net>
* ghostop14 <ghostop14@gmail.com>
* hcab14 <hcab14@gmail.com>
* Johannes Demel <demel@uni-bremen.de>
* Marcus Müller <mueller@kit.edu>
* Nathan West <nathan.west@nrl.navy.mil>
* Nathan West <nathan.west@okstate.edu>

I think we're safe when it comes to kernels.

marcusmueller commented:
> @michaelld from https://github.com/gnuradio/greps/pull/33/files around line 64
> 
> > ### All contributors need to re-submit their code under the new license
> > Whether one submitted code under a CLA or with a DCO, every contributor must re-submit their code under the new license.
> > A contributor may submit their code to any project, including VOLK under a new license, even if that code was submitted under a CLA.
> 
> I assume the correct phrasing would be "I re-submit my code ..."

Slightly confused, as on the call you said you did not want the code to be re-submitted, but instead have a list of authors who are willing to license as LGPL; so you explicitly did not want to start with a clean-slate LGPLv3 repo (or branch) and submit code into it as authors agree. In that case, no resubmission of code can happen, the code is already where it belongs...

So, I think I might have simply misunderstood you there. Do you want to either:

* Start from scratch and only submit code into a clean LGPLv3 codebase (resubmission approach)  xor
* Want to keep a list of authors that agreed to the new license and then flush out the code of which the authors were not willing to relicense?

This PR only makes sense in the second case. So, could you clarify?

jdemel commented:
Yes. This is confusing. The way I understand @bhilburn is, we ask all contributors to re-submit their code under a new license. 
BUT: This is a legal discussion. Technically, they re-submit their code by adding their names to the list.

michaelld commented:
Ditto what @jdemel wrote: "resubmit code" is being used as a legal term when relicensing that for our case will mean adding their name / info to the noted file. It does not mean literally submitting their code again in a PR for us to review (etc) but under the new license.

michaelld commented:
I'm good with moving the license naming in the relicensing file to "LGPL3+" ... I think that's clear enough and precise enough

jdemel commented on COPYING-LGPL:
```diff
@@ -0,0 +1,175 @@
+find below.
+
+However, you'll find some files that carry a license header that assigns them as
+GPLv3, which was the default license for VOLK up to version 3. You can find the
```
Could you change this to `before version 3`? That'd be more verbose. At least for me.

jdemel committed:
relicensing: Add my name to resubmitting authors

With this commit I add my name, to the list of authors who resubmit their
VOLK code contributions under LGPL-3.0-or-later.

I added a few review comments that already came up during the original
PR discussion. Mostly, I rewrote this PR from "granting..." to
"resubmitting...". This is in line with the GREP discussion on the same
subject.
Further, I changed the file extension of
`AUTHORS_RESUBMITTING_UNDER_LGPL.md` from `.txt` to `.md` and fixed an
issue with table formatting.

I, Johannes Demel, Github handle "jdemel", email addresses <ufcsy@student.kit.edu>,
<demel@uni-bremen.de>, <demel@ant.uni-bremen.de>, hereby resubmit all my
contributions to VOLK under LGPL-3.0-or-later.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

mbr0wn commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,22 @@
+|------------|-----------------------------|---------------|--------------------------|
+| 2021-07-01 | Marcus Müller               |               |                          |
+| 2021-08-01 | Johannes Demel              | jdemel        | <ufcsy@student.kit.edu>, <demel@uni-bremen.de>, <demel@ant.uni-bremen.de> |
+|            |                             |
```
```suggestion
| 2021-08-25 | Martin Braun | mbr0wn        | martin@gnuradio.org, martin.braun@ettus.com  |
|            |                             |
```

Ka-zam replied:
Add me:
| 2021-08-27 | Magnus Lundmark | Ka-zam | magnus@skysense.io |

mbr0wn reviewed (commented):
We should not require people to update this file themselves -- if we have an email approval from them, or a github comment approval, we can update this file ourselves.

jdemel committed:
Add Martin Braun

Co-authored-by: Martin Braun <martin@gnuradio.org>
Signed-off-by: Marcus Müller <marcus@hostalia.de>

jdemel commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,23 @@
+this gives the right to use the current contributions under both LGPL-3.0-or-later and that other license.
+Future contributions by these authors are, however, 
+licensed under LGPL-3.0-or-later, unless explicitly stated otherwise.
+
+
```
As per @mbr0wn 's comment, we should add a sentence that tells people how to add themselves to this list.
```suggestion

## How to add myself to the table of authors
There are several options. Choose the one you prefer.
1. Open a PR and add yourself to the table.
2. Add a Github comment that clearly states your wish to be added to the table.
3. Write an email to the maintainers that clearly states your wish to be added to the table.

If you go with option 2 or 3, the maintainers add your statement to the corresponding git commit message for documentation purposes.
```

carlesfernandez committed:
Add Carles Fernandez as re-licesing author

Remove trailing blank spaces

Signed-off-by: Marcus Müller <marcus@hostalia.de>

jdemel committed:
relicensing: Add contributor statement

In https://github.com/gnuradio/volk/pull/486/files/03522eeba80b6d9d40fba7682090c1208872b39f#r697452367
we received:
```
Add me:
| 2021-08-27 | Magnus Lundmark | Ka-zam | magnus@skysense.io |
```

Contributor is added to the list as requested.

jdemel committed:
relicensing: Add how to info

Make it easier to do things right by suggesting specific steps.

mbr0wn commented:
On the re-submit thing: Currently, VOLK2 is a project that is owned by the FSF and licensed GPLv3+. It's not owned by us, so we can't change a license. But we *can* create an entirely new project, VOLK3, that is owned by all, and LGPLv3+. Now, we'll still use the same repo and filenames etc., which also allows us to do `git blame`, `git bisect`, or whatever else we want across 2/3 boundaries. But in principle, we are creating an entirely new project with a new license and ownership. So I think re-submit is the right term, because contributor's can't give us permission the change the license of the existing project.

jdemel reviewed (approved):
LGTM.

Thanks @marcusmueller for starting this.

michaelld commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+# List of authors resubmitting their contributions under LGPL-3.0-or-later
+
+VOLK is going to migrating from GNU General Public License v3.0 or later (GPL-3.0-or-later)
```
"is going to migrating" -> "is migrating" ... or something along those lines

jdemel replied:
```suggestion
VOLK is migrating from GNU General Public License v3.0 or later (GPL-3.0-or-later)
```

michaelld commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+This file is a list of the authors who agreed to resubmit their code
+under the LGPL-3.0-or-later license to this repository.
+In case the affected code is licensed differently than VOLK's current license (GPL-3.0-or-later),
+this gives the right to use the current contributions under both LGPL-3.0-or-later and that other license.
```
"this gives the right" -> "this gives the VOLK project the right"

jdemel replied:
```suggestion
this gives the VOLK project the right to use the current contributions under both LGPL-3.0-or-later and that other license.
```

michaelld commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+2. Add a Github comment that clearly states your wish to be added to the table.
+3. Write an email to the maintainers that clearly states your wish to be added to the table.
+
+If you go with option 2 or 3, the maintainers add your statement to the corresponding git commit message for documentation purposes.
```
"maintainers add your" -> "maintainers will add your"

jdemel replied:
```suggestion
If you go with option 2 or 3, the maintainers will add your statement to the corresponding git commit message for documentation purposes.
```

michaelld commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+
+### Information for option 2 and 3
+
+If you go with option 2 or 3, the maintainers add your statement to the corresponding git commit message for documentation purposes.
```
duplicate sentence to the immediate above one ... choose one or the other

michaelld commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+| 2021-08-01 | Johannes Demel              | jdemel        | <ufcsy@student.kit.edu>, <demel@uni-bremen.de>, <demel@ant.uni-bremen.de> |
+| 2021-08-25 | Martin Braun | mbr0wn        | martin@gnuradio.org, martin.braun@ettus.com  |
+| 2021-08-27 | Carles Fernandez            | carlesfernandez | <carles.fernandez@cttc.es> |
+| 2021-08-27 | Magnus Lundmark | Ka-zam | magnus@skysense.io |
```
Please add mine:

` | 2021-09-04 | Michael Dickens | michaelld | mlk@alum.mit.edu, michael.dickens@ettus.com |`

michaelld commented on COPYING-LGPL:
```diff
@@ -0,0 +1,175 @@
+Files in this code repository are to be licensed under the LGPLv3, which you'll
```
In the files `COPYING` and `README.md` we're using the terms `LGPL-3.0-or-later` and `GPL-3.0-or-later` whereas here we're using `LGPLv3` and `GPLv3`. This usage is inconsistent IMHO. I don't know which is the better / more correct way to go, but let's just be consistent ... but if I had to pick I'd use the terms from the other files.

michaelld commented on README.md:
```diff
@@ -98,23 +98,10 @@ We want to make sure VOLK compiles on a wide variety of compilers. Thus, we targ
+VOLK is moving from the GNU General Public License v3.0 or later (GPL-3.0-or-later) 
+to the GNU Lesser General Public License v3.0 or later (LGPL-3.0-or-later).
+At this point in time, much of the code in the repository is still GPL-licensed, 
+but new contributors are asked to use the LGPLv3+ for their code contributions. 
```
"are asked" -> "are required"

michaelld replied:
`but new contributors` -> `New contributors`

jdemel replied:
```suggestion
New contributors are required to use the LGPLv3+ for their code contributions. 
```

michaelld commented on README.md:
```diff
@@ -98,23 +98,10 @@ We want to make sure VOLK compiles on a wide variety of compilers. Thus, we targ
-> the Free Software Foundation, Inc., 51 Franklin Street,
-> Boston, MA 02110-1301, USA.
->
+VOLK is moving from the GNU General Public License v3.0 or later (GPL-3.0-or-later) 
```
"is moving" -> "is migrating" ... same as the file `AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md`

jdemel replied:
```suggestion
VOLK is migrating from the GNU General Public License v3.0 or later (GPL-3.0-or-later) 
```

michaelld commented on README.md:
```diff
@@ -98,23 +98,10 @@ We want to make sure VOLK compiles on a wide variety of compilers. Thus, we targ
->
+VOLK is moving from the GNU General Public License v3.0 or later (GPL-3.0-or-later) 
+to the GNU Lesser General Public License v3.0 or later (LGPL-3.0-or-later).
+At this point in time, much of the code in the repository is still GPL-licensed, 
```
`is still GPL-licensed,` -> `is still GPL-licensed.`

jdemel replied:
```suggestion
At this point in time, much of the code in the repository is still GPL-licensed.
```

michaelld commented on README.md:
```diff
@@ -98,23 +98,10 @@ We want to make sure VOLK compiles on a wide variety of compilers. Thus, we targ
+to the GNU Lesser General Public License v3.0 or later (LGPL-3.0-or-later).
+At this point in time, much of the code in the repository is still GPL-licensed, 
+but new contributors are asked to use the LGPLv3+ for their code contributions. 
+Existing contributors are very kindly requested to also allow LPGL-licensing by 
```
`to also allow LPGL-licensing by` -> `to relicense their GPL-3.0-or-later license code contributions to LGPL-3.0-or-later by`

jdemel replied:
```suggestion
Existing contributors are very kindly requested to re-submit their GPL-3.0-or-later license code contributions to LGPL-3.0-or-later by
```
After various discussion, we ask everyone to re-submit their code. However, all contributors can re-submit by adding themselves to the list instead of actually opening a new PR with all their commits. That would be infeasible to handle.

michaelld reviewed (changes requested):
Sorry for coming in late to the party ... better late than later?! Great start!

jdemel commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+2. Add a Github comment that clearly states your wish to be added to the table.
+3. Write an email to the maintainers that clearly states your wish to be added to the table.
+
+If you go with option 2 or 3, the maintainers add your statement to the corresponding git commit message for documentation purposes.
```
```suggestion
```

jdemel commented on AUTHORS_RESUBMITTING_UNDER_LGPL_LICENSE.md:
```diff
@@ -0,0 +1,43 @@
+| 2021-08-25 | Martin Braun | mbr0wn        | martin@gnuradio.org, martin.braun@ettus.com  |
+| 2021-08-27 | Carles Fernandez            | carlesfernandez | <carles.fernandez@cttc.es> |
+| 2021-08-27 | Magnus Lundmark | Ka-zam | magnus@skysense.io |
+|            |                             |
```
```suggestion
| 2021-09-04 | Michael Dickens | michaelld | mlk@alum.mit.edu, michael.dickens@ettus.com |
|            |                             |
```

marcusmueller replied:
please don't add more contributor lines to this PR: merge it first, I don't want to become the gatekeeper/bottleneck. Obviously, I'm adding Michael, but he'll be the last one.

jdemel commented:
@michaelld I added suggestions to all of your comments. I really like that feature. It highlights the exact changes and one can go through, "add them to the batch" and just do a commit.
However, I'd separate your addition to the list into its own commit.

marcusmueller committed:
Michael Dickens' remarks

Co-authored-by: Johannes Demel <johannes@demels.de>
Signed-off-by: Marcus Müller <marcus@hostalia.de>

marcusmueller committed:
unified usage of (L)GPL-3.0-or-later

Signed-off-by: Marcus Müller <marcus@hostalia.de>

marcusmueller committed:
Added Michael Dickens

Signed-off-by: Marcus Müller <marcus@hostalia.de>

marcusmueller committed:
unified Email addresses: no <>

Signed-off-by: Marcus Müller <marcus@hostalia.de>

marcusmueller committed:
Marcus Müller Details

Signed-off-by: Marcus Müller <marcus@hostalia.de>

marcusmueller commented:
please don't add more contributor lines to this PR: merge it first, I don't want to become the gatekeeper/bottleneck. Obviously, I'm adding Michael, but he'll be the last one.

marcusmueller commented:
@michaelld @mbr0wn 
Can you check whether your concerns have been addressed? Since @jdemel has already LGTM'ed, that would be the final blocker.

We need to merge this, as I will not be accepting any further PRs-on-PR to add re-submitters to the table anymore.

marcusmueller commented:
@jdemel  as maintainer, you have push access to this branch. Please sign off on your own commits!

michaelld reviewed (approved):
LGMT ; thx for the quick changes per my prior review!

michaelld commented:
@jdemel given the (slight) changes since your last review, I think it's worth the effort for you to re-review. Assuming you approve then please go ahead & merge ... and: Let's do this!!!
